vmbindings/dummyvm/src/api.rs

```diff
@@ -2,24 +2,30 @@
 #![allow(clippy::not_unsafe_ptr_arg_deref)]
 
 use libc::c_char;
+use std::sync::atomic::Ordering;
 use std::ffi::CStr;
 use mmtk::memory_manager;
 use mmtk::AllocationSemantics;
 use mmtk::util::{ObjectReference, Address};
 use mmtk::util::opaque_pointer::*;
 use mmtk::scheduler::{GCController, GCWorker};
 use mmtk::Mutator;
-use mmtk::MMTK;
 use crate::DummyVM;
 use crate::SINGLETON;
+use crate::BUILDER;
 
 #[no_mangle]
-pub extern "C" fn mmtk_gc_init(heap_size: usize) {
-    // # Safety
-    // Casting `SINGLETON` as mutable is safe because `gc_init` will only be executed once by a single thread during startup.
-    #[allow(clippy::cast_ref_to_mut)]
-    let singleton_mut = unsafe { &mut *(&*SINGLETON as *const MMTK<DummyVM> as *mut MMTK<DummyVM>) };
-    memory_manager::gc_init(singleton_mut, heap_size)
+pub extern "C" fn mmtk_init(heap_size: usize) {
+    // set heap size first
+    {
+        let mut builder = BUILDER.lock().unwrap();
+        assert!(builder.options.heap_size.set(heap_size));
```

Comment:
Although Rust always executes code in `assert!`, it is generally not considered a good practice to put code with important side effect inside assertions.